kits/token-wrapper/contracts/TokenWrapper.sol

```diff
@@ -0,0 +1,143 @@
+    string private constant ERROR_CALLER_NOT_TOKEN = "TW_CALLER_NOT_TOKEN";
+    string private constant ERROR_LOCK_AMOUNT_ZERO = "TW_LOCK_AMOUNT_ZERO";
+    string private constant ERROR_INVALID_UNLOCK_AMOUNT = "TW_INVALID_UNLOCK_AMOUNT";
+    string private constant ERROR_INVALID_TOKEN_CONTROLLER = "TW_INVALID_TOKEN_CONTROLLER";
+    string private constant ERROR_TOKEN_BURN_FAILED = "TW_TOKEN_BURN_FAILED";
+    string private constant ERROR_TOKEN_MINT_FAILED = "TW_TOKEN_MINT_FAILED";
+    string private constant ERROR_ERC20_TRANSFER_FAILED = "TW_ERC20_TRANSFER_FAILED";
+    string private constant ERROR_ERC20_TRANSFER_FROM_FAILED = "TW_ERC20_TRANSFER_FROM_FAILED";
+
+    bytes32 public constant CONTROLLER_MANAGER_ROLE = keccak256("CONTROLLER_MANAGER_ROLE");
+
+    ERC20 public erc20;
+    MiniMeToken public token;
+    mapping(address => uint256) internal lockedAmount;
+
+    event TokensLocked(address entity, uint256 amount);
+    event TokensUnlocked(address entity, uint256 amount);
+
+    modifier onlyToken() {
+        require(msg.sender == address(token), ERROR_CALLER_NOT_TOKEN);
+        _;
+    }
+
+    /**
+     * @notice Initialize a new MiniMe token that will be the controller of an ERC20 token
+     * @param _token The MiniMe token that is used in the DAO
+     * @param _erc20 The ERC20 token that is locked in order to receive MiniMe tokens
+     */
+    function initialize(MiniMeToken _token, ERC20 _erc20) external {
+        initialized();
+        require(_token.controller() == address(this), ERROR_INVALID_TOKEN_CONTROLLER);
+
+        token = _token;
+        erc20 = _erc20;
+        token.enableTransfers(false);
+    }
+
+    /**
+     * @notice Lock `_amount` tokens
+     */
+    function lock(uint256 _amount) external {
+        require(_amount > 0, ERROR_LOCK_AMOUNT_ZERO);
+
+        lockedAmount[msg.sender] = lockedAmount[msg.sender].add(_amount);
+        emit TokensLocked(msg.sender, _amount);
+
+        require(erc20.transferFrom(msg.sender, address(this), _amount), ERROR_ERC20_TRANSFER_FROM_FAILED);
+        require(token.generateTokens(msg.sender, _amount), ERROR_TOKEN_MINT_FAILED);
+    }
+
+    /**
+     * @notice Unlock `_amount` tokens
+     */
+    function unlock(uint256 _amount) external {
+        require(_amount <= lockedAmount[msg.sender], ERROR_INVALID_UNLOCK_AMOUNT);
+
+        lockedAmount[msg.sender] = lockedAmount[msg.sender].sub(_amount);
+        emit TokensUnlocked(msg.sender, _amount);
+
+        require(token.destroyTokens(msg.sender, _amount), ERROR_TOKEN_BURN_FAILED);
+        require(erc20.transfer(msg.sender, _amount), ERROR_ERC20_TRANSFER_FAILED);
+    }
+
+    /**
+     * @notice Give control of the MiniMe token over to `_newController`
+     * @param _newController Ethereum address of contract that will control the token
+     */
+    function changeTokenController(address _newController) external auth(CONTROLLER_MANAGER_ROLE) {
```

Comment:
Hm, what's the purpose of changing the controller? If this is run, then `lock` and `unlock` won't work, because `generateTokens` and `destroyTokens` will fail, right?

Reply:
I think it could work if you want to delegate it to another token wrapper instance for example.

Reply:
Hm, yes, but how would you transfer all the locking balances?

Reply:
Yea, it doesn't seem helpful. It was part of the code we started from with @luisivan, maybe he has more context on whether this is actually needed or not.

Reply:
Confirmed that the example with fork from was using it for what I said above along with a token migration strategy, since we removed that, will drop this one too.